frontends/PyCDE/src/pycde/module.py

```diff
@@ -41,135 +42,231 @@ class Input(ModuleDecl):
 class Parameter:
   __slots__ = ["name", "attr"]
 
-  def __init__(self, value, name: str = None):
-    self.attr = support.var_to_attribute(value)
+  def __init__(self, value=None, name: str = None):
+    if value is not None:
+      self.attr = support.var_to_attribute(value)
     self.name = name
 
 
-def module(cls):
+class module:
+  """Decorator for module classes or functions which parameterize module
+  classes."""
+
+  mod = None
+  func = None
+  extern_mod = None
+
+  # When the decorator is attached, this runs.
+  def __init__(self, func_or_class, extern_name=None):
+    self.extern_name = extern_name
+    if inspect.isclass(func_or_class):
+      # If it's just a module class, we should wrap it immediately
+      self.mod = _module_base(func_or_class)
+      _register_generator(self.mod.__name__, "extern_instantiate",
+                          self._instantiate)
+      return
+    elif not inspect.isfunction(func_or_class):
+      raise TypeError("@module got invalid object")
+
+    # If it's a module parameterization function, inspect the arguments to
+    # ensure sanity.
+    self.func = func_or_class
+    self.sig = inspect.signature(self.func)
+    for (_, param) in self.sig.parameters.items():
```

Comment:
nit: i think you could use `parameters.values()` if you don't care about the keys in the dict

Reply:
Yeah, I tried that and it didn't work for some reason. Likely because parameters isn't a standard dict and didn't have the values method, but I don't recall.

Reply:
Oh good point, it is not a standard dict, my bad.